feat(modeling): preserve color for offset and extrude

platypii opened the pull request:
This PR preserves the color attribute when applying operations like `offset` and `extrude` and `project`.

This seems like a clear improvement for users of JSCAD. If you have a 2D object with color like on the left, would you expect extrusion to preserve the color? Or lose it?

![extrudeColor](https://github.com/jscad/OpenJSCAD.org/assets/1766297/275086b8-3fd1-4049-8df1-059db413fe3e)

<details>
  <summary>preserve-color.js</summary>

```js
import * as jscad from '@jscad/modeling'
const { colorize } = jscad.colors
const { extrudeLinear } = jscad.extrusions
const { square } = jscad.primitives
const { translate } = jscad.transforms

export const main = () => {
  const obj2d = [
    colorize([1, 0, 0], square({ center: [0, -4] })),
    colorize([0, 1, 0], square({ center: [0, 0] })),
    colorize([0, 0, 1], square({ center: [0, 4] })),
  ]
  return [
    translate([-4, 0], obj2d),
    translate([4, 0, 0], extrudeLinear({}, obj2d)),
  ]
}
```

</details>

If someone really want no color, it's easy to remove using `colorize`. But if you _want_ to preserve color of sub-components through an operation... there's really no good way to do it without this change.



### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Does your submission pass tests?

hrgdavor reviewed (dismissed):
This a move in right direction. Colors from boolean ops were removed for perf gain, but it is possible we will be able to make keeping colors optional in the future for booleans.

z3dev reviewed (changes requested):
@platypii nice set of changes, and very nice improvements for the user experience.

i just have a bunch of nits. sorry.

platypii commented on packages/modeling/src/operations/extrusions/extrudeRotate.js:
```diff
@@ -59,6 +59,7 @@ export const extrudeRotate = (options, geometry) => {
   // convert geometry to an array of sides, easier to deal with
   let shapeSides = geom2.toSides(geometry)
   if (shapeSides.length === 0) throw new Error('the given geometry cannot be empty')
+  let sliceGeometry = geometry
```
The name change is needed for the end of the function to do:

```
if (geometry.color) output.color = geometry.color
```

But if we overwrite `geometry` at the beginning of the function, then we can't copy the color later.

z3dev replied:
Got it.

platypii commented on packages/modeling/src/operations/extrusions/extrudeRotate.js:
```diff
@@ -59,6 +59,7 @@ export const extrudeRotate = (options, geometry) => {
   // convert geometry to an array of sides, easier to deal with
   let shapeSides = geom2.toSides(geometry)
   if (shapeSides.length === 0) throw new Error('the given geometry cannot be empty')
```
Agreed, I'll make the same change here

platypii commented on packages/modeling/src/operations/offsets/offsetPath2.js:
```diff
@@ -74,17 +74,17 @@ export const offsetPath2 = (options, geometry) => {
   const closed = geometry.isClosed
   const points = path2.toPoints(geometry)
-  if (points.length === 0) throw new Error('the given geometry cannot be empty')
+  if (points.length === 0) return geometry
```
Strongly disagree!

The point of immutable data structures, is that we can return the same empty geometry and be confident that it's okay. If we start making copies of object unnecessarily, it will cost a lot in memory usage and efficiency.

z3dev replied:
Well then, I think that the throw is a better solution.

platypii replied:
No, throw is much worse in my opinion! If you really want me to return a copy I will do that, I really do **not** want to throw.

The geometry is not "incorrect". The empty geometry is a perfectly valid geometry. We should absolutely allow users to `expand` or `offset` or `project` any set of valid geometries, some of which might be empty. If JSCAD throws an exception, you can't perform these operations without painstakingly checking every geometry if it's empty or not.

What does it mean to expand an empty geometry? Empty geometry.
What does it mean to offset an empty geometry? Empty geometry.
What does it mean to extrude an empty geometry? Empty geometry.

Please do not make me throw an error here.

And if not throwing, what is the advantage of making a copy instead of returning the original?

z3dev replied:
I'm not totally disagreeing. I agree that returning a geometry is better choice than throwing an error, but I'm not convinced that returning the original is correct.

There are lots of operations that will accept empty geometries but those operations don't return the original. Consistency is important.

I cannot come up an immediate use case, but if a geometry has a special attribute set then the result should be something else.

I acknowledge that a new object creates an allocation. But users know these are heavy operations. The use of create() is as lightweight as possible.

z3dev replied:
Alternatively, just let the extrude happen on an empty set of points.

platypii commented on packages/modeling/src/operations/offsets/offsetPath2.js:
```diff
@@ -74,17 +74,17 @@ export const offsetPath2 = (options, geometry) => {
-  }
+  const output = geometry.isClosed ?
+    createGeometryFromClosedOffsets(paths) :
+    createGeometryFromExpandedOpenPath(paths, segments, corners, delta)
```
Those existed before this PR. If you have a better suggestion, I am happy to change.

platypii replied:
I just renamed them to slightly better (but still long) names:
`createGeometryFromClosedPath`
`createGeometryFromOpenPath`

platypii commented on packages/modeling/src/geometries/geom2/reverse.js:
```diff
@@ -13,5 +13,6 @@ import { clone } from './clone.js'
 export const reverse = (geometry) => {
   const reversed = clone(geometry)
   reversed.outlines = reversed.outlines.map((outline) => outline.slice().reverse())
```
Will do. I never liked `clone` anyway (#1201)

platypii commented on packages/modeling/src/operations/extrusions/extrudeLinear.test.js:
```diff
@@ -37,6 +39,17 @@ test('extrudeLinear (defaults)', (t) => {
+test('extrudeLinear: preserves color', (t) => {
+  const red = colorize([1, 0, 0], square())
+  const extruded = extrudeLinear({ }, red)
+  t.deepEqual(extruded.color, [1, 0, 0, 1])
```
Sorry I'm not sure what you're suggesting here? Do you want me to also check the color of the `red` square?

z3dev replied:
Ha ha ha... can you change "red" to "redSquare"?

platypii commented on packages/modeling/src/operations/offsets/offsetShell.js:
```diff
@@ -71,6 +71,9 @@ export const offsetShell = (options, geometry) => {
   const { delta, segments } = Object.assign({ }, defaults, options)
 
+  const polygons = geom3.toPolygons(geometry)
+  if (polygons.length === 0) return geometry
```
Same as above. What's the scenario where it helps to return a copy?

platypii commented:
@z3dev I just pushed some more changes which I hope will address your concerns:

 - returns new geometries even if the shape is the same
 - handles empty geometries when it makes sense
 - added tests for the above
 - addressed other PR feedback (function and variable names)

Let me know what you think

z3dev reviewed (approved):
@platypii thanks for hanging in there. Sometimes it take time to sort out differences in opinion.

The changes are all good now, with improvements everywhere.
